pkg/cloud/services/ec2/instances.go

```diff
@@ -319,18 +358,14 @@ func (s *Service) runInstance(role string, i *infrav1.Instance) (*infrav1.Instan
 			return nil, errors.Wrap(err, "failed to decode bootstrapData")
 		}
 
-		gz := gzip.NewWriter(&buf)
-		if _, err := gz.Write(decoded); err != nil {
-			return nil, errors.Wrap(err, "failed to gzip userdata")
-		}
-
-		if err := gz.Close(); err != nil {
-			return nil, errors.Wrap(err, "failed to gzip userdata")
+		compressed, err := zipBytes(decoded)
```

Comment:
@detiber it looks like we're encoding & decoding & reencoding the user data multiple times? Would it make sense to leave everything unencoded until we zip it & are then ready to encode?